lib/chef/context.rb

```diff
@@ -0,0 +1,104 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require_relative "licensing_config"
+
+class Chef
+  class Context
+    class << self
+      # The test kitchen will generate a nonce and sign it with a secret key.
+      # All these details are written to a file in the temp directory with the same name as the secret key.
+      # The file looks like:
+      # nonce:<nonce>
+      # timestamp:<timestamp>
+      # signature:<signature>
+      #
+      # The nonce is valid for 60 seconds from the time it was generated.
+      # The secret key will be passed to chef-infra client through env variable.
+      #
+      # This method reads the file and verifies the nonce, timestamp and signature.
+      # If the received timestamp is within 60 seconds of the current time and the signature is valid
+      # this confirms that the execution is initiated from the test-kitchen.
+      def test_kitchen_context?
+        return @context unless @context.nil?
+
+        @context = false
+        return @context if context_secret.nil? || context_secret.empty?
+
+        if File.exist?(signed_file_path)
+          # Read the nonce, timestamp and signature from the file
+          received_nonce, received_timestamp, received_signature = read_file_content
+          current_time = Time.now.utc.to_i
+          # Check if the nonce is within 60 seconds of the current time
+          if (current_time - received_timestamp.to_i).abs < 60
+            if expected_signature(received_nonce, received_timestamp) == received_signature
+              @context = true
+            end
+          end
+          # Delete the file after reading the content
+          File.delete(signed_file_path)
+        end
+
+        @context
+      end
+
+      # This method will switch the license entitlement to Chef Workstation entitlement.
+      def switch_to_workstation_entitlement
+        puts "Running under Test-Kitchen: Switching License to Chef Workstation entitlement"
+        ChefLicensing.configure do |config|
+          # Reset entitlement ID to the ID of Chef Workstation
+          config.chef_entitlement_id = Chef::LicensingConfig::WORKSTATION_ENTITLEMENT_ID
+        end
+      end
+
+      private
+
+      # The secret key is passed as an environment variable to the chef-infra client.
+      def context_secret
+        ENV.fetch("TEST_KITCHEN_CONTEXT", "")
+      end
+
+      # The file contains the nonce, timestamp and signature which are written by the test-kitchen.
+      def signed_file_path
+        "#{Dir.tmpdir}/kitchen/#{context_secret}"
+      end
+
+      # Reads the file and return the nonce, timestamp and signature
+      def read_file_content
+        file_content = {}
+        File.open(signed_file_path, "r:bom|utf-16le:utf-8") do |file|
```

Comment:
Do you have a reference for this syntax? It took a bit of digging to refresh my memory. Also, is this specifically due to Windows?

Reply:
Yeah, there are some encoding issues on Windows and I had to add the encoding to fix it. When the test-kitchen creates the files on the node, it causes some encoding issues.

Reply:
@ashiqueps Even though it's a slim possibility of issue, it still feels safer to check if `windows?` first:

```ruby
def self.signed_file_encoding
  @signed_file_encoding ||= windows? "r:bom|utf-16le:utf-8" : "r"
end
```

and then
```ruby
File.open(signed_file_path, Context.signed_file_path) do |file|
```